### DISCUSSION TITLE
Is there a way to disable auto deploy on specific branch?

### DISCUSSION THREAD

**Discussion by siloneco** [vercel/vercel#4307] [General] [answered]
I'm sorry if this has already been asked by others.
I'm developing with 3 branches on private repository, but they all get deployed on all branches and quickly become rate-limited.

Is there a way to configure a particular branch not to deploy automatically?

If it doesn't, I'll give up, so it's no problem. sorry for my bad English. thx.

**Comment by coetry** [comment]
Hi @siloneco there's no way to turn off deploys _per branch_, but you can turn off the Git integration altogether with the following [config option](https://vercel.com/docs/configuration#git-integrations/github-enabled)

```json
{
  "github": {
    "enabled": false
  }
}
```
If you have three branches named A, B, and C and you only push to B, then only a single deployment should be made for the B branch. If you push to B one time and 3 deployments are made, then there is a bug. Can you confirm that only a single deployment is being made for a single push?

**Reply by paymog** [reply]
Is this still relevant? The linked docs don't seem to exist anymore.

**Reply by utybo** [reply]
Seems like the relevant docs have been moved here: https://vercel.com/docs/project-configuration#git-configuration/github-enabled

**Reply by amyegan** [reply]
Thanks for sharing the updated link. There is also a new [`git.deploymentEnabled`](https://vercel.com/docs/projects/project-configuration/git-configuration#git.deploymentenabled) option that lets you disable auto-deployment on individual branches.

**Reply by jacob-mako** [reply]
I think this is the working link: https://vercel.com/docs/concepts/projects/project-configuration/git-configuration

**Reply by muizidn** [reply]
I am writing this post to help everyone who want to disable the auto deployment. https://www.mozzlog.com/blog/how-to-disable-vercel-auto-deploy-for-git

**Comment by siloneco** [comment]
Thank you for quick reply. I tried it in the develop branch and waited a minute after pushing it, and it didn't do anything. I think I can do something with this feature. Thank you so much!

**Comment by allisson-lima** [comment]
In file vercel.json

{
  "git": {
    "deploymentEnabled": {
      "main": false
    }
  }
}

**Reply by amyegan** [reply]
Thanks for sharing this answer! Using the new [`git.deploymentEnabled`](https://vercel.com/docs/project-configuration#git-configuration/git-deployment-enabled) option is a perfect solution for disabling a specific branch.

**Reply by stavgafny** [reply]
Yes, using `deploymentEnabled` you can now disable a specific branch. [See docs](https://vercel.com/docs/concepts/projects/project-configuration/git-configuration#git.deploymentenabled)

**Reply by coreyward** [reply]
So if I want to disable all branches other than my main branch, I would just add 1.20892582e64 pre-generated combinations of 40 character strings to my vercel.json? 🤨

**Comment by coreyward** [comment]
Is there still no way to only trigger preview builds for pull requests? Netlify has offered this for an eternity. I don't need Vercel running doomed builds for branches that are not ready to at least open a PR for.

**Comment by sambacha** [comment]
> Is there still no way to only trigger preview builds for pull requests? Netlify has offered this for an eternity. I don't need Vercel running doomed builds for branches that are not ready to at least open a PR for.

You can enable the "Ignored Build Step" field by referring to the documentation of this feature. If the command returns "0", the build will be skipped. If, however, a code "1" or greater is returned, then a new deployment will be built.

So bash script with 
~~~bash
exit 1;
~~~

Will let the build proceed. 



This is quite honestly absurd, you have to wonder how much billable hours vercel gets out of this, the feature is obviously easy to implement considering the existing git integration.

**Reply by coreyward** [reply]
To be honest, I don't know what any of this means. Ignored build steps? Where do I write said shell script? Why is it a shell script and not, um, JS? How does my shell script get access to the Github repo to know whether or not a branch is involved in a pull request?

> This is quite honestly absurd, you have to wonder how much billable hours vercel gets out of this, the feature is obviously easy to implement considering the existing git integration.

I think you're onto something here.

**Reply by leerob** [reply]
Please see the latest selected answer - this is possible without needing an ignored build step!

**Reply by coreyward** [reply]
@leerob I want to only run builds on pull requests and the `main` branch. There doesn't seem to be any way of doing that.

**Comment by leerob** [comment, marked as answer]
Yes, you can disable deployments for specific branches through `vercel.json`:

```
{
  "git": {
    "deploymentEnabled": {
      "my-branch-name": false
    }
  }
}
```

https://vercel.com/docs/projects/project-configuration/git-configuration

**Reply by coreyward** [reply]
The opposite would be far more useful: an option to only run builds on specified branches. But we still also need a way for Vercel to only run builds on pull requests, too. Building arbitrary branches that are for speculative work is not useful, and asking devs to update `vercel.json` to add the current branch each time they create one is unrealistic.

**Reply by NickG-NZ** [reply]
Agreed, absurd that the default is to build for every push to every branch.. It should be vastly simpler to configure to only build on push (or merge request) to specific branches.
Yes we can configure which branches build on push with a bash script (with practically zero documentation), but come on...

**Reply by hwrdtm** [reply]
Can we use patterns (eg. `release-*`) for the keys under `git.deploymentEnabled`?

**Comment by jciurc** [comment]
How do I disable it for every branch except main?

**Reply by amyegan** [reply]
If you want to prevent all but **production** deployments, you can use [Ignored Build Step](https://vercel.com/docs/projects/overview#ignored-build-step) and select the "**only build production**" option. Or you can use a customized script to allow only specified branches, like
```bash
#!/bin/bash

echo "VERCEL_GIT_COMMIT_REF: $VERCEL_GIT_COMMIT_REF"

if [[ "$VERCEL_GIT_COMMIT_REF" == "main" || "$VERCEL_GIT_COMMIT_REF" == "staging"  ]] ; then
  # Proceed with the build
    echo "✅ - Build can proceed"
  exit 1;

else
  # Don't build
  echo "🛑 - Build cancelled"
  exit 0;
fi
```

https://vercel.com/guides/how-do-i-use-the-ignored-build-step-field-on-vercel
